reactive/charm.py

```diff
@@ -7,332 +7,287 @@
-    revision = ""
-    if os.path.exists(version_file):
-        with open(version_file, encoding="utf-8") as f:
+class SMTPRelayCharm(ops.CharmBase):
+    """SMTP Relay."""
+
+    def __init__(self, *args: Any) -> None:
+        """SMTP Relay."""
+        super().__init__(*args)
+
+        self.framework.observe(self.on.upgrade_charm, self._reconcile)
+        self.framework.observe(self.on.install, self._reconcile)
+        self.framework.observe(self.on.config_changed, self._reconcile)
+        self.framework.observe(self.on.peer_relation_changed, self._reconcile)
+        self.framework.observe(self.on.milter_relation_changed, self._reconcile)
+
+    def _reconcile(self, _: ops.EventBase) -> None:
+        self.unit.status = ops.MaintenanceStatus("Reconciling SMTP relay")
+        charm_state = State.from_charm(self.config)
+
+        self._install()
+        self._configure_smtp_auth(charm_state)
+        self._configure_smtp_relay(charm_state)
+        self._configure_policyd_spf(charm_state)
+        self._set_active()
+
+    @staticmethod
+    def _install(logrotate_conf_path: str = "/etc/logrotate.d/rsyslog") -> None:
+        """Configure logging."""
+        utils.copy_file("files/fgrepmail-logs.py", "/usr/local/bin/fgrepmail-logs", perms=0o755)
+        utils.copy_file("files/50-default.conf", "/etc/rsyslog.d/50-default.conf", perms=0o644)
+        contents = utils.update_logrotate_conf(logrotate_conf_path)
+        utils.write_file(contents, logrotate_conf_path)
+
+    def _configure_smtp_auth(
+        self,
+        charm_state: State,
+        dovecot_config: str = "/etc/dovecot/dovecot.conf",
+        dovecot_users: str = "/etc/dovecot/users",
+    ) -> None:
+        """Ensure SMTP authentication is configured or disabled via Dovecot."""
+        self.unit.status = ops.MaintenanceStatus("Setting up SMTP authentication (dovecot)")
+
+        contents = construct_dovecot_config_file_content(
+            dovecot_users, charm_state.enable_smtp_auth
+        )
+        changed = utils.write_file(contents, dovecot_config)
+
+        if charm_state.smtp_auth_users:
+            contents = construct_dovecot_user_file_content(charm_state.smtp_auth_users)
+            utils.write_file(contents, dovecot_users, perms=0o640, group="dovecot")
+
+        if not charm_state.enable_smtp_auth:
+            self.unit.status = ops.MaintenanceStatus(
+                "SMTP authentication not enabled, ensuring ports are closed"
+            )
+            self.unit.close_port("tcp", 465)
+            self.unit.close_port("tcp", 587)
+            systemd.service_stop("dovecot")
+            # XXX: mask systemd service disable
+            return
+
+        self.unit.status = ops.MaintenanceStatus(
+            "Opening additional ports for SMTP authentication"
+        )
+        self.unit.open_port("tcp", 465)
+        self.unit.open_port("tcp", 587)
+
+        if changed:
```

Comment:
Reload will falback to restart if stopped, so this is not needed:
https://github.com/canonical/smtp-relay-operator/pull/40/files#diff-e70f91cc11fc9ab3f2c1ed313b235a3a93b41547a49f6b9641302af4b34f790cR210